File: apps/api/app/routers/plan_route.py
"""POST /api/plan-route — geo itinerary + optional Claude tips."""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, HTTPException
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field

from app.constants.regions import REGION_COORDINATES, REGION_DB_ID
from app.services.live_route_candidates import load_live_route_candidates
from app.services.plan_itinerary import build_skeleton, enrich_with_claude

router = APIRouter(tags=["route"])

REGION_LABELS: dict[str, str] = {
    "quba": "Quba",
    "qusar": "Qusar",
    "seki": "Şəki",
    "sheki": "Şəki",
    "qabala": "Qəbələ",
    "gabala": "Qəbələ",
    "lerik": "Lerik",
    "baku": "Bakı",
}


class WeatherIn(BaseModel):
    prefer_indoor: bool = False
    summary_az: str | None = None
    exclude_categories: list[str] = Field(default_factory=list)
    prefer_categories: list[str] = Field(default_factory=list)


class PoisIn(BaseModel):
    restaurants: list[dict[str, Any]] = Field(default_factory=list)
    accommodations: list[dict[str, Any]] = Field(default_factory=list)
    attractions: list[dict[str, Any]] = Field(default_factory=list)


class PlanRouteIn(BaseModel):
    region: str
    days: int = Field(ge=1, le=7)
    budget: str = "mid"
    interests: list[str] = Field(default_factory=list)
    groupType: str | None = "solo"
    weather: WeatherIn | None = None
    pois: PoisIn | None = None
    # Travel-from-home (e.g. Baku → Quba day trip)
    fromOrigin: bool = False
    originLat: float | None = None
    originLng: float | None = None
    departTime: str | None = "08:00"
    returnByTime: str | None = "21:00"


@router.post("/api/plan-route")
def plan_route_endpoint(body: PlanRouteIn) -> JSONResponse:
    region_key = body.region.strip().lower()
    if region_key not in REGION_COORDINATES:
        raise HTTPException(
            status_code=400,
            detail={
                "error": "Invalid region",
                "allowed_regions": list(REGION_COORDINATES.keys()),
            },
        )

    db_region = REGION_DB_ID.get(region_key, region_key)
    region_label = REGION_LABELS.get(region_key) or REGION_LABELS.get(db_region) or db_region
    interests = [str(i) for i in (body.interests or [])]
    candidate_source = "client"

    if body.pois and (
        body.pois.restaurants or body.pois.accommodations or body.pois.attractions
    ):
        restaurants = list(body.pois.restaurants)
        accommodations = list(body.pois.accommodations)
        attractions = list(body.pois.attractions)
    else:
        loaded = load_live_route_candidates(
            region_key,
            per_bucket=12,
            interests=interests or None,
            source="google",
        )
        buckets = loaded["buckets"]
        restaurants = buckets["restaurants"]
        accommodations = buckets["accommodations"]
        attractions = buckets["attractions"]
        candidate_source = str(loaded.get("source") or "google")

    if not (restaurants or accommodations or attractions):
        raise HTTPException(
            status_code=400,
            detail={"error": "Bu bölgədə yer tapılmadı"},
        )

    weather = body.weather.model_dump() if body.weather else None
    group_type = (body.groupType or "solo").strip() or "solo"

    try:
        skeleton = build_skeleton(
            region=region_key,
            days=int(body.days),
            budget=body.budget,
            interests=interests,
            group_type=group_type,
            restaurants=restaurants,
            accommodations=accommodations,
            attractions=attractions,
            weather=weather,
            origin_lat=body.originLat,
            origin_lng=body.originLng,
            from_origin=bool(body.fromOrigin),
            depart_time=body.departTime or "08:00",
            return_by_time=body.returnByTime or "21:00",
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail={"error": str(exc)}) from exc

    plan = enrich_with_claude(
        skeleton,
        region_label=region_label,
        days=int(body.days),
        budget=body.budget,
        interests=interests,
        group_type=group_type,
        weather=weather,
    )

    travel = plan.pop("travel", None) or skeleton.get("travel")
    lodging = plan.pop("lodging", None) or skeleton.get("lodging")
    plan.pop("meta", None)

    return JSONResponse(
        content={
            "success": True,
            "summary": plan.get("summary"),
            "days": plan.get("days"),
            "total_cost": plan.get("total_cost"),
            "best_time": plan.get("best_time") or skeleton.get("best_time"),
            "region": db_region,
            "regionLabel": region_label,
            "travel": travel,
            "lodging": lodging,
            "source": "fastapi_geo",
            "candidatesSource": candidate_source,
        }
    )

File: apps/mobile/lib/planRoute.ts
/**
 * Plan AI route via FastAPI only (single source of truth).
 * Edge Function algorithm is deprecated — no second planner.
 * See docs/ARCHITECTURE.md → "AI marşrut (plan-route)".
 */

import { getApiBaseUrl } from './apiBase';
import { trackEvent } from './trackEvent';

export type PlanRouteStop = {
  poi_id?: string;
  id?: string;
  name: string;
  category: string;
  lat: number;
  lng: number;
  time?: string;
  duration?: string;
  duration_minutes?: number;
  tip?: string;
  daypart?: string;
};

export type PlanRouteDay = {
  day: number;
  title: string;
  stops: PlanRouteStop[];
  estimated_cost?: string;
  notes?: string;
};

export type PlanRouteResult = {
  summary: string;
  days: PlanRouteDay[];
  total_cost?: string;
  best_time?: string;
  region?: string;
  regionLabel?: string;
  source?: string;
  travel?: {
    from_origin?: boolean;
    outbound_minutes?: number;
    return_minutes?: number;
    depart_origin_at?: string;
    arrive_region_at?: string;
    leave_region_by?: string;
    return_origin_by?: string;
    distance_km?: number;
  } | null;
  lodging?: {
    id?: string;
    name?: string;
    category?: string;
    lat?: number;
    lng?: number;
    note?: string;
  } | null;
};

export type PlanRouteWeather = {
  prefer_indoor: boolean;
  summary_az?: string;
  exclude_categories?: string[];
  prefer_categories?: string[];
};

export type PlanRoutePois = {
  restaurants: unknown[];
  accommodations: unknown[];
  attractions: unknown[];
};

export type PlanRouteInput = {
  region: string;
  days: number;
  budget: string;
  interests: string[];
  groupType?: string;
  weather?: PlanRouteWeather | null;
  pois?: PlanRoutePois;
  fromOrigin?: boolean;
  originLat?: number | null;
  originLng?: number | null;
  departTime?: string;
  returnByTime?: string;
  /** Changes on each "plan again" so near-tied stops vary */
  varietySeed?: number;
  /** Soft-exclude POIs from the previous plan */
  excludePoiIds?: string[];
};

async function planRouteViaFastApi(
  input: PlanRouteInput
): Promise<PlanRouteResult | null> {
  const base = getApiBaseUrl();
  if (!base) {
    return null;
  }

  try {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), 60_000);
    const res = await fetch(`${base}/api/plan-route`, {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        region: input.region,
        days: input.days,
        budget: input.budget,
        interests: input.interests,
        groupType: input.groupType ?? 'solo',
        weather: input.weather ?? null,
        pois: input.pois ?? null,
        fromOrigin: Boolean(input.fromOrigin),
        originLat: input.originLat ?? null,
        originLng: input.originLng ?? null,
        departTime: input.departTime ?? '08:00',
        returnByTime: input.returnByTime ?? '21:00',
        varietySeed: input.varietySeed ?? Date.now(),
        excludePoiIds: input.excludePoiIds ?? [],
      }),
      signal: controller.signal,
    });
    clearTimeout(timer);

    const data = (await res.json()) as PlanRouteResult & {
      success?: boolean;
      error?: string;
      detail?: { error?: string } | string;
    };

    if (!res.ok) {
      const detail =
        typeof data.detail === 'object' && data.detail?.error
          ? data.detail.error
          : typeof data.detail === 'string'
            ? data.detail
            : data.error;
      // Business errors: do not fall back to Edge
      if (res.status >= 400 && res.status < 500) {
        throw new Error(detail || `Plan API xətası (${res.status})`);
      }
      return null;
    }

    if (!data?.days || !Array.isArray(data.days)) {
      return null;
    }

    return {
      summary: data.summary ?? 'Marşrut hazırlandı.',
      days: data.days.map((day) => ({
        ...day,
        stops: Array.isArray(day.stops) ? day.stops : [],
      })),
      total_cost: data.total_cost,
      best_time: data.best_time,
      region: data.region,
      regionLabel: data.regionLabel,
      travel: (data as PlanRouteResult).travel ?? null,
      lodging: (data as PlanRouteResult).lodging ?? null,
      source: data.source ?? 'fastapi_geo',
    };
  } catch (err) {
    if (err instanceof Error) {
      // Propagate 4xx business errors
      if (
        err.message.includes('yer tapılmadı') ||
        err.message.includes('kifayət') ||
        err.message.includes('Invalid region') ||
        err.message.includes('Plan API xətası (4')
      ) {
        throw err;
      }
    }
    return null;
  }
}

/** FastAPI geo planner only — retry once on network/5xx; no Edge algorithm. */
export async function planRoute(input: PlanRouteInput): Promise<PlanRouteResult> {
  const first = await planRouteViaFastApi(input);
  if (first) {
    void trackEvent('plan_route_success', {
      region: input.region,
      days: input.days,
      source: first.source ?? 'fastapi',
    });
    return first;
  }
  // Brief retry for transient Railway/network blips
  await new Promise((r) => setTimeout(r, 600));
  const second = await planRouteViaFastApi(input);
  if (second) {
    void trackEvent('plan_route_success', {
      region: input.region,
      days: input.days,
      source: second.source ?? 'fastapi_retry',
    });
    return second;
  }
  throw new Error(
    'Marşrut serveri əlçatan deyil. İnterneti yoxlayın və yenidən cəhd edin.'
  );
}
